File: Scripts/Python/Solucoes_analiticas_da_barras.py
import numpy as np
import matplotlib.pyplot as plt

# Entradas
L = 2.0                   # Comprimento da barra [m]
nx = 100                  # Número de pontos em x
x = np.linspace(0, L, nx) # Coordenadas x
h = L / (nx - 1)          # Comprimento dos intervalos
A = 0.01                  # Área da seção [m^2]
E = 210e9                 # Módulo de Young [Pa]
# Tipos de função: 'con', 'lin', 'par', 'sin', 'cos', 'exp'
type = 'con'
q0 = 1e6                  # Carga distribuída [N/m]
ua = 0                    # Deslocamento à esquerda [m]
ub = 0                    # Deslocamento à direita [m]

# Definições da carga
y = x / L                 # Vetor de posição normalizada
if type == 'con':          # Constante
    p = q0 * np.ones(len(y))
elif type == 'lin':        # Linear
    p = q0 * y
elif type == 'par':        # Parábola
    p = q0 * y * (1 - y)
elif type == 'sin':        # Seno
    p = q0 * np.sin(np.pi * y)
elif type == 'cos':        # Cosseno
    p = q0 * np.cos(np.pi * y)
elif type == 'exp':        # Exponencial
    p = q0 * np.exp(y)
else:                      # Padrão = constante
    print(f'Tipo de carregamento >>> {type} <<< não implementado !!!')

# Obtenha a solução particular up e sua derivada
if type == 'con':          # Constante
    up = -q0 * x**2 / (2 * E * A)
    dup = -q0 * x / (E * A)
elif type == 'lin':        # Linear
    up = -q0 * x**3 / (6 * E * A * L)
    dup = -q0 * x**2 / (2 * E * A * L)
elif type == 'par':        # Parábola
    up = (-q0 * x**3 * (2 * L - x)) / (12 * E * A * L**2)
    dup = (-q0 * x**2 * (3 * L - 2 * x)) / (6 * E * A * L**2)
elif type == 'sin':        # Seno
    up = q0 * (L / np.pi)**2 * np.sin(np.pi * x / L) / (E * A)
    dup = q0 * (L / np.pi) * np.cos(np.pi * x / L) / (E * A)
elif type == 'cos':        # Cosseno
    up = q0 * (L / np.pi)**2 * np.cos(np.pi * x / L) / (E * A)
    dup = -q0 * (L / np.pi) * np.sin(np.pi * x / L) / (E * A)
elif type == 'exp':        # Exponencial
    up = -q0 * L**2 * np.exp(x / L) / (E * A)
    dup = -q0 * L * np.exp(x / L) / (E * A)
else:                      # Padrão = constante
    print(f'Tipo de solução >>> {type} <<< não implementado !!!')

# Obtenha os valores constantes para uh e sua derivada duh
up0 = up[0]
upL = up[-1]
c2 = ua - up0
c1 = (ub - upL - c2) / L

# Obtenha a solução uana e sua derivada duana
uana = up + c1 * x + c2
duana = dup + c1

# Plotar o carregamento distribuído
plt.subplot(2, 1, 1)
plt.bar(x, p)
plt.grid(True)
plt.title('Carregamento Distribuído')
plt.xlabel('Posição [m]')
plt.ylabel('Carga [N/m]')

# Plotar o campo de deslocamento e sua derivada
plt.subplot(2, 1, 2)
plt.plot(x, uana, 'r', label='u(x)')
plt.plot(x, duana, 'g', label='du(x)/dx')
plt.grid(True)
plt.title('Campo de deslocamento e sua derivada')
plt.xlabel('Posição [m]')
plt.ylabel('Deslocamento [m]')
plt.legend()

plt.tight_layout()
plt.show()

File: Scripts/MATLAB/Solucoes_analiticas_da_barras.m
clear all; close all; clc;

% Data set
L = 2.0;                  % bar length [m]
nx = 100;                 % number of x-points
x = linspace(0, L, nx)';   % x coordinates
h = L/(nx-1);             % intervals length
A = 0.01;                 % section area [m^2]
E = 210.E9;               % Young modulus [Pa]
% function types: con, lin, par, sin, cos, exp
type = 'con';
q0 = 1E6;                 % Distributed Load [N/m]
ua = 0;                   % left displacement [m]
ub = 0;                   % right displacement [m]

% Load definitions
y = x/L;                  % normalized position vector
switch type
    case 'con' % constant
        p = q0*ones(length(y),1);
    case 'lin' % linear
        p = q0*y;
    case 'par' % parabola
        p = q0*y.*(1-y);
    case 'sin' % sinus
        p = q0*sin(pi*y);
    case 'cos' % cosinus
        p = q0*cos(pi*y);
    case 'exp' % exponential
        p = q0*exp(y);
    otherwise  % default = constant
        warning(['tipo de carregamento >>> ',type,' <<< nao implementado !!!'])
end

% get particular solution up and its derivative
switch type
case 'con' % constant
  up = -q0*x.^2/(2*E*A);
  dup = -q0*x/(E*A);
case 'lin' % linear
  up = -q0*x.^3/(6*E*A*L);
  dup = -q0*x.^2/(2*E*A*L);
case 'par' % parabola
  up = (-q0*x.^3.*(2*L-x))./(12*E*A*L^2);
  dup = (-q0*x.^2.*(3*L-2*x))./(6*E*A*L^2);
case 'sin' % sinus
  up = q0*(L/pi)^2*sin(pi*x/L)/(E*A);
  dup = q0*(L/pi)*cos(pi*x/L)/(E*A);
case 'cos' % cosinus
  up = q0*(L/pi)^2*cos(pi*x/L)/(E*A);
  dup = -q0*(L/pi)*sin(pi*x/L)/(E*A);
case 'exp' % exponential
  up = -q0*L^2*exp(x/L)/(E*A);
  dup = -q0*L*exp(x/L)/(E*A);
otherwise % default = constant
  warning(['tipo de carreamento >>>', type, '<<<<< nao implementado !!!'])
end
% get Constant Values for uh and its derivative duh
up0 = up(1);
upL = up(end);
c2 = ua - up0;
c1 = (ub - upL - c2)/L;
% get solution uana and its derivative duana
uana = up + c1*x+c2;
duana = dup + c1;
subplot(2,1,1), bar(x,p);
grid on;
title('Carregamento Distribuido')
xlabel('posicao [m]')
ylabel('carga [N/m]')
subplot(2,1,2), plot(x, uana, 'r',x, duana,'g'); grid on;
title ('Campo de deslocamento e sua derivada')

xlabel('posicao [m]')
ylabel('Deslocamentos [m]')
legend('u(x)', 'du(x)/dx')